app/services/applicable_measure_unit_merger.rb

```diff
@@ -14,29 +18,35 @@ def call
   private
 
   def delta_measure_units
-    clean_unhandled_measure_units do
-      uk_filtered_commodity.applicable_measure_units.merge(xi_filtered_commodity.applicable_measure_units)
-    end
+    applicable_units =  uk_filtered_commodity.applicable_measure_units.merge(xi_filtered_commodity.applicable_measure_units)
+
+    clean_and_dedup_measure_units(applicable_units)
   end
 
   def uk_measure_units
-    clean_unhandled_measure_units do
-      uk_filtered_commodity.applicable_measure_units
-    end
+    applicable_units = uk_filtered_commodity.applicable_measure_units
+
+    clean_and_dedup_measure_units(applicable_units)
   end
 
   def xi_measure_units
-    clean_unhandled_measure_units do
-      xi_filtered_commodity.applicable_measure_units.merge(uk_filtered_commodity.applicable_excise_measure_units)
-    end
+    applicable_units = xi_filtered_commodity.applicable_measure_units.merge(uk_filtered_commodity.applicable_excise_measure_units)
+
+    clean_and_dedup_measure_units(applicable_units)
   end
 
-  def clean_unhandled_measure_units
-    yield.tap do |units|
-      UNHANDLED_MEASURE_UNITS.each do |unhandled_unit|
-        units.delete(unhandled_unit)
+  def clean_and_dedup_measure_units(applicable_units)
+    if @dedupe
+      applicable_units = applicable_units.each_with_object({}) do |(measurement_unit_code, measurement_unit_values), acc|
+        acc[measurement_unit_values['coerced_measurement_unit_code'].presence || measurement_unit_code] = measurement_unit_values
       end
     end
+
+    UNHANDLED_MEASURE_UNITS.each do |unhandled_unit|
```

Comment:
`applicable_units.without(UNHANDLED_MEASURE_UNITS)` ?

Reply:
I think except would do it, yeah.